File: src/template/pedro/Operacion.java
package template.pedro;

abstract class Operacion {
    protected double primerValor;
    protected double segundoValor;
    protected double resultado;
    protected Integer redondeo = null;

    final void ejecuta(double primerValor, double segundoValor) {
        getOperacion();
        setValores(primerValor, segundoValor);
        calcula();
        muestraResultado();
    }

    abstract void getOperacion();
    abstract void calcula();

    final void setValores(double primerValor, double segundoValor) {
        this.primerValor = primerValor;
        this.segundoValor = segundoValor;
        System.out.println("primer valor: " + this.primerValor);
        System.out.println("segundo valor: " + this.segundoValor);
    }

    final void muestraResultado() {
        if (redondeo != null && redondeo >= 0) {
            System.out.println("redondeo a " + redondeo + " decimales");
            System.out.println("resultado: " + String.format("%." + redondeo + "f", resultado));
        } else {
            System.out.println("sin redondear");
            System.out.println("resultado: " + resultado);
        }

        System.out.println("\n");
    }
}

File: manuel/template/Operacion.php
<?php

    abstract class Operacion{
        protected $valor1;
        protected $valor2;
        protected $resultado;

        final function ejecuta($valor1, $valor2)
        {
            $this->getOperacion();
            $this->setValores($valor1, $valor2);
            $this->calcula();
            $this->muestraResultado();
            echo "<br/>";
        }

        protected function setValores($valor1, $valor2)
        {
            $this->valor1 = $valor1;
            $this->valor2 = $valor2;
        }
        
        protected function muestraResultado()
        {
            echo "El resultado es: $this->resultado <br>";
        }
        
        protected function enFraccion()
        {
            return false;
        }

        abstract function calcula();
        abstract function getOperacion();
    }

?>